torch/lib/THD/base/data_channels/Store.cpp

```diff
@@ -0,0 +1,205 @@
+  }
+  // after sending the query, wait for a 'stop_waiting' reponse
+  recv_bytes<query_type>(socket, &query, 1);
+  if (query != query_type::stop_waiting)
+    throw std::runtime_error("stop_waiting response expected");
+}
+
+static void oracle_daemon() {
+  port_type port;
+  rank_type world_size;
+  int socket;
+  std::tie(port, world_size) = load_master_env();
+  std::vector<int> sockets(world_size);
+  std::tie(socket, std::ignore) = listen(port);
+  // accept WORLD_SIZE connections
+  for (int i = 0; i < world_size; i++) {
+    int p_socket;
+    std::string p_addr;
+    rank_type p_rank;
+    std::tie(p_socket, p_addr) = accept(socket);
+    recv_bytes<rank_type>(p_socket, &p_rank, 1);
+    sockets[p_rank] = p_socket;
+  }
+
+  // listen for requests
+  struct pollfd fds[world_size];
+  int timeout_msecs = -1;
+  for (int i = 0; i < world_size; i++) {
+    fds[i].fd = sockets[i];
+    fds[i].events = POLLIN | POLLOUT;   // we only read or write
+  }
+
+  // receive the queries
+  bool finished = false;
+  while (!finished) {
+    for (int i = 0; i < world_size; i++) fds[i].revents = 0;
+    SYSCHECK(poll(fds, world_size, timeout_msecs));
+    for (int i = 0; i < world_size; i++) {
+      if ((fds[i].revents & POLLIN) == POLLIN) {
+        finished = query(fds[i].fd);
+      }
+      if ((fds[i].revents & POLLERR) == POLLERR) {
+        throw std::runtime_error("problem in poll");
+      }
+    }
+  }
+}
+
+/* 
+ * query communicates with the worker. The format
+ * of the query is as follows:
+ * type of query | size of arg1 | arg1 | size of arg2 | arg2 | ...
+ * or, in the case of wait
+ * type of query | number of args | size of arg1 | arg1 | ...
+ */
+static bool query(int socket) {
+  static store_type store_;
+  static std::unordered_map<int, std::vector<std::string>> waiting;
+  query_type qt;
+  recv_bytes<query_type>(socket, &qt, 1);
+  if (qt == set) {
+    size_type key_size;
+    recv_bytes<size_type>(socket, &key_size, 1);
+    std::vector<char> key_raw(key_size);
+    recv_bytes<char>(socket, key_raw.data(), key_raw.size());
+    size_type data_size;
+    recv_bytes<size_type>(socket, &data_size, 1);
+    std::vector<char> data(data_size);
+    recv_bytes<char>(socket, data.data(), data.size());
+    store_[std::string(key_raw.data(), key_raw.size())] = std::move(data);
+    // On "set" wake up all of the processes that wait
+    // for keys already in the store
+    for (auto& process : waiting) {
+      auto& p_socket = process.first;
+      auto& p_keys = process.second;
+      if (check(store_, p_keys))
+          wakeup(p_socket);
+    }
+    return false;
+  } else if (qt == get) {
+    size_type key_size;
+    size_type data_size;
+    recv_bytes<size_type>(socket, &key_size, 1);
+    std::vector<char> key_raw(key_size);
+    recv_bytes<char>(socket, key_raw.data(), key_raw.size());
+    auto& data = store_.at(std::string(key_raw.data(), key_raw.size()));
```

Comment:
I think that this should be an exception - `get` is always being done after `wait`